src/minter-contract-utils/build.rs

```diff
@@ -88,3 +115,121 @@ fn get_solidity_contract<'a>(
+    times_file_path: PathBuf,
+    solidity_output_dir: PathBuf,
+}
+
+impl SolidityFileTimesWatcher {
+    /// Creates a new SolidityFileTimesWatcher
+    pub fn new(subpaths: &'static [&'static str]) -> Self {
+        let root_path =
+            solidity_helper::workspace_root_path().expect("Failed to get workspace root path");
+
+        let solidity_root_path = solidity_helper::solidity_root_path(&root_path, None)
+            .expect("Failed to get solidity root path");
+
+        let subpaths = subpaths
+            .iter()
+            .map(|subpath| solidity_root_path.join(subpath))
+            .collect::<Vec<_>>();
+
+        let times_file_path = root_path.join("target/.solidity_file_timestamps.toml");
+
+        let solidity_output_dir = root_path
+            .join("target/")
+            .join(format!("{}", rand::random::<u64>()));
+
+        Self {
+            subpaths,
+            times_file_path,
+            solidity_output_dir,
+        }
+    }
+
+    /// Checks whether the files have changed since the last time they were checked
+    pub fn have_files_changed(&self, current_file_times: Option<&SolidityFileTimes>) -> bool {
+        let current_file_times = match current_file_times {
+            Some(file_times) => file_times,
+            None => {
+                println!("Current file times is None. Assuming files have changed");
+                return true;
+            }
+        };
+        let new_file_times = self.scan_file_times().expect("Failed to scan file times");
+        // first check whether there are new files
+        if current_file_times.files.len() != new_file_times.files.len() {
+            return true;
+        }
+
+        // check whether the files have changed
+        current_file_times
+            .files
+            .keys()
+            .chain(new_file_times.files.keys())
+            .any(|key| current_file_times.files.get(key) != new_file_times.files.get(key))
+    }
+
+    /// Saves the file times to the times file
+    pub fn save_file_times(&self) -> anyhow::Result<()> {
+        let file_times = self.scan_file_times()?;
+        let file_times = toml::to_string(&file_times)?;
+
+        let mut file = std::fs::File::create(&self.times_file_path)?;
+        file.write_all(file_times.as_bytes())?;
+
+        Ok(())
+    }
+
+    fn load_file_times(&self) -> anyhow::Result<SolidityFileTimes> {
+        let file_times = std::fs::read_to_string(&self.times_file_path)?;
+
+        Ok(toml::from_str(&file_times)?)
+    }
+
+    fn scan_file_times(&self) -> anyhow::Result<SolidityFileTimes> {
+        let mut file_times = SolidityFileTimes::new(self.solidity_output_dir.clone());
+
+        for subpath in &self.subpaths {
+            Self::scan_dir_file_times(&mut file_times, subpath)?;
+        }
+
+        Ok(file_times)
+    }
+
+    fn scan_dir_file_times(files: &mut SolidityFileTimes, path: &Path) -> anyhow::Result<()> {
+        for file_in_dir in std::fs::read_dir(path)? {
```

Comment:
is this recursive? `walkdir` is a common option to simplify this type of operations